schemamd/behaviors.go

```diff
@@ -4,68 +4,65 @@ import (
 	tfjson "github.com/hashicorp/terraform-json"
 )
 
-// childIsRequired returns true for blocks with min items > 0 or explicitly required
-// attributes
-func childIsRequired(block *tfjson.SchemaBlockType, att *tfjson.SchemaAttribute) bool {
-	if att != nil {
-		return att.Required
-	}
+func childAttributeIsRequired(att *tfjson.SchemaAttribute) bool {
+	return att.Required
+}
 
+func childBlockIsRequired(block *tfjson.SchemaBlockType) bool {
 	return block.MinItems > 0
 }
 
-// childIsOptional returns true for blocks with with min items 0, but any required or
-// optional children, or explicitly optional attributes
-func childIsOptional(block *tfjson.SchemaBlockType, att *tfjson.SchemaAttribute) bool {
-	if att != nil {
-		return att.Optional
-	}
+func childAttributeIsOptional(att *tfjson.SchemaAttribute) bool {
+	return att.Optional
+}
 
+// childBlockIsOptional returns true for blocks with with min items 0 and any required or optional children.
+func childBlockIsOptional(block *tfjson.SchemaBlockType) bool {
 	if block.MinItems > 0 {
 		return false
 	}
 
 	for _, childBlock := range block.Block.NestedBlocks {
-		if childIsRequired(childBlock, nil) {
+		if childBlockIsRequired(childBlock) {
 			return true
 		}
-		if childIsOptional(childBlock, nil) {
+		if childBlockIsOptional(childBlock) {
 			return true
 		}
 	}
 
 	for _, childAtt := range block.Block.Attributes {
-		if childIsRequired(nil, childAtt) {
+		if childAttributeIsRequired(childAtt) {
 			return true
 		}
-		if childIsOptional(nil, childAtt) {
+		if childAttributeIsOptional(childAtt) {
 			return true
 		}
 	}
 
 	return false
 }
 
-// childIsReadOnly returns true for blocks where all leaves are read only (computed
-// but not optional)
-func childIsReadOnly(block *tfjson.SchemaBlockType, att *tfjson.SchemaAttribute) bool {
-	if att != nil {
-		// these shouldn't be able to be required, but just in case
-		return att.Computed && !att.Optional && !att.Required
-	}
+// Read-only is computed but not optional.
+func childAttributeIsReadOnly(att *tfjson.SchemaAttribute) bool {
+	// these shouldn't be able to be required, but just in case
+	return att.Computed && !att.Optional && !att.Required
+}
 
+// childBlockIsReadOnly returns true for blocks where all leaves are read-only.
```

Comment:
I know you didn't introduce this bug, but just noting that this may be buggy: an empty block being set vs an empty block being null _could_ still be semantically different, meaning I don't know that I would classify it as "read-only".